Make the packager work with babel strict mode transform

janicduplessis opened the pull request:
At the moment we have to disable strict mode for the transform-es2015-modules-commonjs because strict mode leaks to the global scope and breaks the bridge. It was due to the way the polyfills were bundled in the package. To fix it, I wrapped the polyfill modules in an IIFE. Then when strict mode was enabled some polyfills were broken due to strict mode errors so that was fixed too. Also removed the IIFE from the polyfills that included one.

This diff doesn't enable the strict mode transform since some internal facebook modules depend on it not being enabled. When #5214 lands we could make the default babel config shipped with OSS react-native use strict mode modules and facebook could just modify the babel config to disable it if needed.

This will allow removing `"strict": false` from https://github.com/facebook/react-native/blob/master/packager/react-packager/.babelrc#L16

Fixes #5316

janicduplessis commented on Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js:
```diff
@@ -25,11 +25,11 @@
 
 if (typeof GLOBAL === 'undefined') {
-  GLOBAL = this;
+  global.GLOBAL = this;
```
Should we get rid of the `GLOBAL` global variable and just use the node standard `global` everywhere? Except for this file it is used in 1 or 2 other places in the codebase.

kittens replied:
`GLOBAL` is valid in Node so removing it will probably break some third party npm modules.

janicduplessis replied:
Ok let's leave it like this.

ide commented on packager/react-packager/src/Resolver/polyfills/Array.prototype.es6.js:
```diff
@@ -5,54 +5,51 @@
  * @polyfill
```
Files with `@polyfill` don't get wrapped in a closure scope (i.e. they aren't modules) in the internal Facebook packager... @martinbigio is it OK to make the changes below, in React Native's source?

davidaurelio replied:
We don’t read that annotation anywhere. We just add all polyfills as special cases

davidaurelio replied:
i.e. it should be fine given the changes in `packager/react-packager/src/Resolver/index.js`

vjeux replied:
@ide means www for "internal package manager". This is where @polyfill is being read. I'd love to avoid to fork them qs much as possible

davidaurelio replied:
Ah I see. Does that mean you are against this specific change?

janicduplessis commented:
On a completely unrelated note that new @facebook-github-bot picture is amazing!

satya164 commented:
This is awesome. I was planning to work on this, but seems now I don't have to :D

Thanks for working on this.

satya164 commented on Libraries/JavaScriptAppEngine/polyfills/document.js:
```diff
@@ -1,34 +1,32 @@
-   * The document must be shimmed before anything else that might define the
-   * `ExecutionEnvironment` module (which checks for `document.createElement`).
-   */
+/* eslint strict: 0 */
```
Why `eslint strict: 0`?

janicduplessis replied:
`global-strict` was replaced with `strict` in eslint 1.0 so it was causing some lint warnings.

satya164 commented on packager/react-packager/src/Resolver/__tests__/Resolver-test.js:
```diff
@@ -1020,5 +1028,26 @@ describe('Resolver', function() {
       });
     });
+
+    pit('should resolve polyfills', function () {
```
+1 for adding tests

davidaurelio commented on Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js:
```diff
@@ -70,6 +70,15 @@ function polyfillGlobal(name, newValue, scope = GLOBAL) {
+ */
+function polyfillIfNeeded(name, polyfill, scope = GLOBAL) {
+  if (typeof scope[name] === 'undefined') {
+    Object.defineProperty(scope, name, {value: polyfill});
```
I wonder whether this is generally safe. In browsers, many properties are enumerable/writable/configurable, e.g. `navigator`.

janicduplessis replied:
I added a parameter to configure the descriptor and used it to make navigator enumerable/writable/configurable. For the Number polyfill the properties are all not enumerable/writable/configurable so that is fine.

davidaurelio commented on Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js:
```diff
@@ -70,6 +70,15 @@ function polyfillGlobal(name, newValue, scope = GLOBAL) {
+ * Polyfill a module if it is not already defined in `scope`.
+ */
+function polyfillIfNeeded(name, polyfill, scope = GLOBAL) {
+  if (typeof scope[name] === 'undefined') {
```
no need for `typeof`

davidaurelio commented on Libraries/JavaScriptAppEngine/polyfills/document.js:
```diff
@@ -1,34 +1,32 @@
+/**
+ * The document must be shimmed before anything else that might define the
+ * `ExecutionEnvironment` module (which checks for `document.createElement`).
+ */
```
Do we have a problem with a specific version of the `ExecutionEnvironment` module?

https://github.com/facebook/fbjs/blob/master/src/core/ExecutionEnvironment.js#L14-L18 seems to do the right thing, i.e. it doesn’t strictly need `document` to exist.

vjeux replied:
Correct, all those shouldn't be needed anymore now that we are in a web worker

davidaurelio replied:
:+1: 

will clean this up in a follow up – @janicduplessis, can you just remove the comment for now, please? It’s misleading

janicduplessis replied:
@davidaurelio Should I add something like `// TODO: Remove document polyfill now that chrome debugging is in a web worker`?

davidaurelio replied:
No, that’s fine, but thanks!

davidaurelio commented on Libraries/JavaScriptAppEngine/polyfills/document.js:
```diff
@@ -1,34 +1,32 @@
-      'React element. You probably forgot to require ' + name + '.'
-    );
+// The browser defines Text and Image globals by default. If you forget to
+// require them, then the error message is very confusing.
```
The same is probably true here @amasad @vjeux

davidaurelio commented on Libraries/JavaScriptAppEngine/polyfills/document.js:
```diff
@@ -1,34 +1,32 @@
-})(this);
+// There is no DOM so MutationObserver doesn't make sense. It is used
+// as feature detection in Bluebird Promise implementation
+global.MutationObserver = undefined;
```
and here?

davidaurelio commented on packager/react-packager/src/Resolver/index.js:
```diff
@@ -239,4 +241,12 @@ function defineModuleCode(moduleName, code) {
+function definePolyfillCode(code) {
+  return [
+    '(function(global) {\n',
+    `  ${code}`,
```
Don’t bother with formatting/indentation, it won’t look good with multiline code anyway :stuck_out_tongue:

davidaurelio commented on packager/react-packager/src/Resolver/polyfills/prelude.js:
```diff
@@ -1,5 +1,4 @@
-/* eslint global-strict:0 */
-__DEV__ = false;
+/* eslint strict:0 */
+global.__DEV__ = false;
```
We have a special transform which picks up this variable – I will check whether it still works

vjeux replied:
Maybe jest needs it?

davidaurelio replied:
cc @cpojer 

@vjeux we have the WPO transform that reads it – https://github.com/facebook/react-native/blob/f546dc3790edd4e0b310caccdc273f31a717258c/packager/react-packager/src/transforms/whole-program-optimisations/dead-module-elimination.js#L63-L76
It will break.

davidaurelio commented:
@janicduplessis thank you for the effort! I will try to **shipit** when I’m in the office.

davidaurelio commented:
@janicduplessis, thank you for the effort. I’m generally fine with the changes. Let’s clarify the open questions, especially in polyfills/document.js

davidaurelio commented on packager/react-packager/src/Resolver/index.js:
```diff
@@ -239,4 +241,12 @@ function defineModuleCode(moduleName, code) {
 
+function definePolyfillCode(code) {
+  return [
+    '(function(global) {\n',
```
Please remove the line break here, otherwise line numbers in redboxes will be off by one for polyfills.

janicduplessis commented:
@davidaurelio Thanks for the review, I fixed the issues concerning the navigator polyfill and the other minor stuff. The issues left are the `@polyfill` thing and WPO. I tested WPO on master vs my branch and it is effectively broken (`__DEV__` code is not being removed). I'll check what has to change to make it work.

davidaurelio commented:
Awesome, thank you! Let me know if you need any help

janicduplessis commented:
@davidaurelio Should I keep support for both `__DEV__` and `global.__DEV__`?

davidaurelio commented:
@janicduplessis Thank you, looks good!

davidaurelio commented:
@facebook-github-bot shipit

davidaurelio commented:
@janicduplessis I’m sorry, but you have to rebase :-(

janicduplessis commented:
@davidaurelio Rebased on master and squashed the commits.

davidaurelio commented:
I tried to land immediately after you posted yesterday, and right now, but there are still conflicts in `packager/react-packager/src/Resolver/polyfills/babelHelpers.js` and `packager/react-packager/src/Resolver/polyfills/error-guard.js`.

Afaik there is no need to squash your commits, btw. They will be squashed by our system anyway, so don’t worry doing it.

janicduplessis commented:
@davidaurelio Hmm weird, right now it says the branch has no conflicts. If you want to fix the conflicts manually what I did for these 2 files is simply take the version from master and then remove the extra closure scope that is not needed anymore.

janicduplessis commented:
It seems like you still have the version before I rebased since that's the 2 conflicts I fixed, not sure how your system works exactly but maybe try reimporting it since I rebased it may have trouble updating my branch.

davidaurelio commented:
ok, it’s supposed to update itself. Let’s try to ship from here.

davidaurelio commented:
@facebook-github-bot shipit

davidaurelio commented:
ok, the patch succeeded. Unless any internal tests fail this PR should get closed soon :-)

satya164 commented:
:tada:

corbt commented:
:+1: * :100:
